apps_download/models/product_product.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+
+import os
+import tempfile
+import shutil
+import logging
+import base64
+import subprocess
+import time
+from odoo.exceptions import ValidationError
+from odoo import models, fields, api, _
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'prto_validateoduct_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char('Module Path')
+
+    @api.constrains('dependent_product_ids')
+    def check_dependent_recursion(self):
```

Comment:
[Models contain a `_check_recursion` helper method](https://books.google.com/books?id=Kv7fDAAAQBAJ&pg=PA79&lpg=PA79&dq=odoo+check_recursion&source=bl&ots=jHPk1ymnnO&sig=T7pPui5yecoXcffxTJ_UOHN1y6M&hl=en&sa=X&ved=0ahUKEwjX4ML399TWAhXkhFQKHU1vDZwQ6AEINDAB#v=onepage&q=odoo%20check_recursion&f=false). Is there a reason we are performing the test manually here instead?

Reply:
Yes, because _check_recursion from odoo models it is based on O2m field and we are checking for M2m. Don't know anything specifically about checking recursively for M2m.

Reply:
What about [`_check_m2m_recursion`]? 😉  

Admittedly I didn't know it existed until I looked at the `_check_recursion` code to validate today's TIL

Reply:
Hi !
Same here I also did not know about that feature/Method exists.
Changes are done replace my method with the ORM method.